src/Media/Audio/OpenALSoundProvider.cpp

```diff
@@ -680,6 +681,116 @@ PAudioTrack CreateAudioTrack(const std::string &file_path) {
+    virtual size_t GetChannelCount() { return _baseDataSource->GetChannelCount(); }
+    virtual std::shared_ptr<Blob> GetNextBuffer() { return _baseDataSource->GetNextBuffer(); }
+
+    bool linkSource(ALuint al_source);
+
+ protected:
+    PAudioDataSource _baseDataSource;
+    std::vector<ALuint> _buffers;
+};
+
+OpenALAudioDataSource::~OpenALAudioDataSource() {
+    _baseDataSource->Close();
+    if (_buffers.size()) {
+        alDeleteBuffers(_buffers.size(), &_buffers.front());
+    }
+}
+
+bool OpenALAudioDataSource::Open() {
+    if (_buffers.size()) {
+        return true;
+    }
+
+    _baseDataSource->Open();
+
+    bool result = true;
+    ALsizei al_sample_rate = GetSampleRate();
+    ALenum al_format = AL_FORMAT_STEREO16;
+    unsigned int num_channels = GetChannelCount();
+
+    switch (num_channels) {
+        case 1:
+            al_format = AL_FORMAT_MONO16;
+            break;
+        case 2:
+            al_format = AL_FORMAT_STEREO16;
+            break;
+        default:
+            if (alIsExtensionPresent("AL_EXT_MCFORMATS")) {
+                switch (num_channels) {
+                    case 4:
+                        al_format = alGetEnumValue("AL_FORMAT_QUAD16");
+                        break;
+                    case 6:
+                        al_format = alGetEnumValue("AL_FORMAT_51CHN16");
+                        break;
+                    case 7:
+                        al_format = alGetEnumValue("AL_FORMAT_61CHN16");
+                        break;
+                    case 8:
+                        al_format = alGetEnumValue("AL_FORMAT_71CHN16");
+                        break;
+                }
+            }
+            Error("Unsupported number of audio channels: %u", num_channels);
+    }
+
+    while (true) {
+        std::shared_ptr<Blob> buffer = _baseDataSource->GetNextBuffer();
+        if (!buffer) {
+            break;
+        }
+
+        ALuint al_buffer = -1;
+        alGenBuffers(1, &al_buffer);
+        if (CheckError()) {
+            result = false;
+            break;
+        }
+
+        alBufferData(al_buffer, al_format, buffer->data(), buffer->size(), al_sample_rate);
+        if (CheckError()) {
+            result = false;
+            break;
+        }
+
+        _buffers.push_back(al_buffer);
+    }
+
+    _baseDataSource->Close();
+
+    return result;
+}
+
+void OpenALAudioDataSource::Close() {
+    _baseDataSource->Close();
+}
+
+bool OpenALAudioDataSource::linkSource(ALuint al_source) {
+    alSourceQueueBuffers(al_source, _buffers.size(), &_buffers.front());
+    if (CheckError()) {
+        return false;
+    }
+    return true;
+}
+
+PAudioDataSource PlatformDataSourceInitialize(PAudioDataSource baseDataSource) {
+    std::shared_ptr<OpenALAudioDataSource> source = std::make_shared<OpenALAudioDataSource>(baseDataSource);
+
+    return std::dynamic_pointer_cast<IAudioDataSource, OpenALAudioDataSource>(source);
```

Comment:
You can just do `return std::make_shared` here.

Also I would suggest returning `std::shared_ptr<OpenALAudioDataSource>`.